Fix Merlin configuration server in a non-default build context

voodoos opened the pull request:
This PR fixes #4125.

The workspace initialization was wrong in the `dune ocaml-merlin` binary.
I added a test which is disabled by default as it requires a global switch `default` to exist.

voodoos committed:
Add test for opam contexts

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>

rgrinberg commented:
It should be possible to get the tests to work by default. Have a look at the workspace tests for multi contexts tests.

voodoos commented:
> It should be possible to get the tests to work by default. Have a look at the workspace tests for multi contexts tests.

Looking at other tests (grep `(opam `) I only found two:
- One (`workspaces.t/opma`) is disabled, saying 
> A workspace context can be defined using an opam switch. This test is disabled
because we don't really have a way to mock an opam switch.
- And the other uses a switch named `default` so dune's test-suite already assumes such a switch is present on the system. I will use that switch then.

voodoos committed:
Disable log creation

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>

voodoos committed:
Add CHANGES entry

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>

jeremiedimino commented on bin/ocaml_merlin.ml:
```diff
@@ -28,14 +28,12 @@ let term =
-      Dune_rules.Workspace.init ();
-      ( match dump_config with
+      let open Fiber.O in
+      let+ _workspace = Import.Main.scan_workspace common in
```
This is going to scan the workspace and load all the dune files. This can be quite expensive in a large workspace. It feels like the `dune ocaml-merlin` command shouldn't do this. WDYT?

voodoos replied:
I agree that `dune ocaml-merlin` should not do that.

I there a more lightweight way to detect the workspace configured for Merlin ?

jeremiedimino replied:
This should work: `(Workspace.workspace ()).merlin_context`

voodoos replied:
This is what was already done: 

https://github.com/ocaml/dune/blob/78707ffa288bc45dc61de05af481e7ec935f586b/src/dune_rules/merlin_server.ml#L51-L54

However without the previous call to `scan_workspace` this seems to always return `default`, and that was the reason the issue was opened in the first place.

jeremiedimino replied:
Was it loading the right workspace file?

voodoos commented:
I cherry picked the commit from https://github.com/ocaml/dune/pull/4132 adding the possibility to specify which variant of the `default` context is used for Merlin configuration.

voodoos committed:
Use safer default workspace init.

In the absence of a specified dune-workspace file, always check for one witht the standard name.

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>

voodoos committed:
Pass params to workspace init if present

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>

voodoos committed:
Test update

Signed-off-by: Ulysse Gérard <thevoodoos@gmail.com>

jeremiedimino commented on src/dune_rules/context.ml:
```diff
@@ -743,9 +743,10 @@ let instantiate_context env (workspace : Workspace.t)
       } ->
     let merlin =
-      workspace.merlin_context = Some (Workspace.Context.name context)
+      workspace.merlin_context = Some (Workspace.Context.name context) || merlin
```
Do we still need this? IIRC, the code that parses the workspace file in workspace.ml ensures that there is exactly one context that has `merlin` set to true.

voodoos replied:
Yes it is still needed because the merlin context, when chosen by default, is not reflected in the context `merlin` field, only in `workspace.merlin_context`.

However the `|| merlin` part of the test is superfluous as if the user chose a context to enable merlin this should already be checked by the `workspace.merlin_context` test.

jeremiedimino commented:
> I cherry picked the commit from #4132 adding the possibility to specify which variant of the default context is used for Merlin configuration.

I don't think we should do this. It is going to make it more difficult to backport only the bugfix in 2.8. We could backport both the bugfix and the new feature, but that is going to break our usual compatibility rules. Let's be strict here and add the `(merlin)` field in 2.9. If we want this feature soon, we can just release 2.9 soon.

jeremiedimino commented on test/blackbox-tests/test-cases/merlin/github4125.t/run.t:
```diff
@@ -0,0 +1,39 @@
+  $ unset INSIDE_DUNE
+  $ dune build
```
Something doesn't seem right here. If you unset `INSIDE_DUNE`, then `dune build` is going to:

- read the user configuration file
- lookup the root outside of the test
- and a few other things

Which is going to make it difficult to keep this test reproducible.

Why do you need to unset `INSIDE_DUNE`?

voodoos replied:
Because without it the Dune in the test file doesn't tries to reach the root of the test project and thus the test where I launch `dune ocaml-merlin` from a subfolder fails because the `dune-workspace` file at the root of the project is not loaded. It is a test about root discovery...

With `INDISE_DUNE` unset in the test dune does find the correct root for the test, that is `$TESTCASE_ROOT` and not the root of the real dune project.

Other tests already being at the root pass without problem so i can just remove this one if you prefer.

https://github.com/ocaml/dune/blob/bebb7a9432b149fce95ac820757216372afb8f0f/bin/workspace_root.ml#L90-L93

jeremiedimino replied:
> With INSIDE_DUNE unset in the test dune does find the correct root for the test, that is $TESTCASE_ROOT and not the root of the real dune project.

This is because `dune-workspace` has precedence over `dune-project`. If you add a `dune-workspace` file at the root of dune, then the test will pick this as root.

If we want to test this properly, we need a way to tell Dune to stop the search at `$TESTCASE_ROOT`. But unsetting `INSIDE_DUNE` seems too dodgy.

voodoos replied:
For now I removed the test in question.

voodoos commented:
> > I cherry picked the commit from #4132 adding the possibility to specify which variant of the default context is used for Merlin configuration.
> 
> I don't think we should do this. It is going to make it more difficult to backport only the bugfix in 2.8. We could backport both the bugfix and the new feature, but that is going to break our usual compatibility rules. Let's be strict here and add the `(merlin)` field in 2.9. If we want this feature soon, we can just release 2.9 soon.

You are right, I dropped it.

jeremiedimino commented on CHANGES.md:
```diff
@@ -4,6 +4,9 @@ Unreleased
   @rgrinberg)
 
+- Fixed wrong workspace discovery from `dune ocaml-merlin` and allow custom
+  context based on default to be used for merlin. (#4127, fixes #4125, @voodoos)
```
Seems like the last commit should be reverted

voodoos replied:
I forgot to drop it when rebasing, good catch.

jeremiedimino reviewed (approved):
Looks good to merge once changelog is updated

voodoos commented on test/blackbox-tests/test-cases/merlin/github4125.t/dune-workspace:
```diff
@@ -0,0 +1,9 @@
+ (opam
+  (name cross)
+  (switch default)
+  (merlin)))
```
Sorry for that, I thought other tests were already relying on such a switch but they must have been disabled since.
@rgrinberg said it might be possible to test this kind of feature but I did not found examples of such things in the test-base.

I think disabling this test is fine if don't find a workaround.
